File: frontend/src/lib/simulator/policies.ts
// src/lib/simulator/policies.ts
import type { CacheSet } from './cache';

export interface ReplacementPolicy {
  name: string;
  chooseVictim(cset: CacheSet): number;
  onAccess(cset: CacheSet, way: number, tick: number): void;
  onFill(cset: CacheSet, way: number, tick: number): void;
}

export class LRUPolicy implements ReplacementPolicy {
  name = 'LRU';

  chooseVictim(cset: CacheSet): number {
    let minWay = 0;
    let minTick = Infinity;
    for (let i = 0; i < cset.lines.length; i++) {
      if (cset.lines[i].lastUsed < minTick) {
        minTick = cset.lines[i].lastUsed;
        minWay = i;
      }
    }
    return minWay;
  }

  onAccess(cset: CacheSet, way: number, tick: number): void {
    cset.lines[way].lastUsed = tick;
  }

  onFill(cset: CacheSet, way: number, tick: number): void {
    cset.lines[way].lastUsed = tick;
  }
}

export class FIFOPolicy implements ReplacementPolicy {
  name = 'FIFO';

  chooseVictim(cset: CacheSet): number {
    let minWay = 0;
    let minOrder = Infinity;
    for (let i = 0; i < cset.lines.length; i++) {
      if (cset.lines[i].insertOrder < minOrder) {
        minOrder = cset.lines[i].insertOrder;
        minWay = i;
      }
    }
    return minWay;
  }

  onAccess(_cset: CacheSet, _way: number, _tick: number): void {
    // FIFO doesn't change on access
  }

  onFill(cset: CacheSet, way: number, tick: number): void {
    cset.lines[way].insertOrder = tick;
  }
}

export class LFUPolicy implements ReplacementPolicy {
  name = 'LFU';

  chooseVictim(cset: CacheSet): number {
    let minWay = 0;
    let minFreq = Infinity;
    let minTick = Infinity;

    for (let i = 0; i < cset.lines.length; i++) {
      const line = cset.lines[i];
      if (line.frequency < minFreq) {
        minFreq = line.frequency;
        minTick = line.lastUsed;
        minWay = i;
      } else if (line.frequency === minFreq && line.lastUsed < minTick) {
        // Tie-breaker: LRU
        minTick = line.lastUsed;
        minWay = i;
      }
    }
    return minWay;
  }

  onAccess(cset: CacheSet, way: number, tick: number): void {
    cset.lines[way].frequency++;
    cset.lines[way].lastUsed = tick;
  }

  onFill(cset: CacheSet, way: number, tick: number): void {
    cset.lines[way].frequency = 1;
    cset.lines[way].lastUsed = tick;
  }
}

export function getPolicy(name: string): ReplacementPolicy {
  switch (name.toUpperCase()) {
    case 'LRU': return new LRUPolicy();
    case 'FIFO': return new FIFOPolicy();
    case 'LFU': return new LFUPolicy();
    default: throw new Error(`Unknown policy: ${name}`);
  }
}

File: backend/simulator/policies.py
"""
policies.py - Cache replacement policy implementations.
Supports LRU, FIFO, and LFU strategies.
Each policy operates on a CacheSet and returns the index of the line to evict.
"""
from __future__ import annotations
from typing import Protocol
from .cache import CacheSet


class ReplacementPolicy(Protocol):
    """Interface that all replacement policies must satisfy."""
    name: str

    def choose_victim(self, cset: CacheSet) -> int:
        """Return the index of the cache line to evict."""
        ...

    def on_access(self, cset: CacheSet, way: int, tick: int) -> None:
        """Called on every cache hit or fill to update metadata."""
        ...

    def on_fill(self, cset: CacheSet, way: int, tick: int) -> None:
        """Called specifically on a cache fill (miss that loads data)."""
        ...


class LRUPolicy:
    """Least-Recently-Used: evict the line with the smallest last_used tick."""
    name = "LRU"

    def choose_victim(self, cset: CacheSet) -> int:
        return min(range(len(cset.lines)), key=lambda i: cset.lines[i].last_used)

    def on_access(self, cset: CacheSet, way: int, tick: int) -> None:
        cset.lines[way].last_used = tick

    def on_fill(self, cset: CacheSet, way: int, tick: int) -> None:
        cset.lines[way].last_used = tick


class FIFOPolicy:
    """First-In-First-Out: evict the line with the smallest load_order."""
    name = "FIFO"

    def choose_victim(self, cset: CacheSet) -> int:
        return min(range(len(cset.lines)), key=lambda i: cset.lines[i].load_order)

    def on_access(self, cset: CacheSet, way: int, tick: int) -> None:
        # FIFO does not update metadata on access — order is fixed at load time.
        pass

    def on_fill(self, cset: CacheSet, way: int, tick: int) -> None:
        cset.lines[way].load_order = tick


class LFUPolicy:
    """Least-Frequently-Used: evict the line with the lowest access frequency.
    Ties are broken by least-recently-used (smallest last_used).
    """
    name = "LFU"

    def choose_victim(self, cset: CacheSet) -> int:
        return min(
            range(len(cset.lines)),
            key=lambda i: (cset.lines[i].frequency, cset.lines[i].last_used),
        )

    def on_access(self, cset: CacheSet, way: int, tick: int) -> None:
        cset.lines[way].frequency += 1
        cset.lines[way].last_used = tick

    def on_fill(self, cset: CacheSet, way: int, tick: int) -> None:
        cset.lines[way].frequency = 1
        cset.lines[way].last_used = tick


POLICIES = {
    "LRU": LRUPolicy(),
    "FIFO": FIFOPolicy(),
    "LFU": LFUPolicy(),
}


def get_policy(name: str) -> LRUPolicy | FIFOPolicy | LFUPolicy:
    """Return a policy instance by name (case-insensitive)."""
    key = name.upper()
    if key not in POLICIES:
        raise ValueError(f"Unknown policy '{name}'. Choose from: {list(POLICIES.keys())}")
    return POLICIES[key]
